lib/events.js

```diff
@@ -240,23 +243,25 @@ EventEmitter.prototype.removeListener =
 
 EventEmitter.prototype.removeAllListeners =
     function removeAllListeners(type) {
-      var key, listeners;
+      var listeners, events;
 
-      if (!this._events)
+      if (!(events = this._events))
         return this;
 
       // not listening for removeListener, no need to emit
-      if (!this._events.removeListener) {
+      if (!events.removeListener) {
         if (arguments.length === 0)
           this._events = {};
-        else if (this._events[type])
-          delete this._events[type];
+        else if (events[type])
+          delete events[type];
         return this;
       }
 
       // emit removeListener for all listeners on all events
       if (arguments.length === 0) {
-        for (key in this._events) {
+        var keys = Object.keys(events);
```

Comment:
In fact not only is the for-in itself slow for normalized objects, the entire containing function becomes unoptimizable:

```js
var obj = {key: 3};

delete obj.key;
obj.key2 = 3;
function containsBadForIn(o) {
    for (var key in o) {
        console.log(key);
    }
    return o;
}
containsBadForIn(obj);
containsBadForIn(obj);

%OptimizeFunctionOnNextCall(containsBadForIn);

containsBadForIn(obj);
```

```
$ /c/etc/iojs --allow-natives-syntax --trace-opt js.js
[disabled optimization for 000000008CBF0579 <SharedFunctionInfo containsBadForIn>, reason: ForInStatement is not fast case]
...
```